src/ocamldep.ml

```diff
@@ -99,18 +99,49 @@ module Dep_graphs = struct
                   )
               )
         }
-    (* implementations don't introduce interface deps b/c they don't have
-       interfaces *)
+    (* implementations introduce interface dependencies for private modules
+       only *)
     ; intf =
-        { vlib.intf with
-          per_module =
-            Module.Name.Map.map vlib.intf.per_module ~f:(fun v ->
-              v >>^ List.map ~f:Module.remove_files
-            )
+        { Dep_graph.
+          dir = impl.intf.dir
+        ; per_module =
+            Module.Name.Map.merge vlib.intf.per_module impl.intf.per_module
+              ~f:(fun name vlib impl ->
+                match vlib, impl with
+                | None, None -> assert false
+                | Some v, Some i ->
+                  (* This can happen when we are implementing the virtul module
+                     or this is a private module. In the former case, we should
+                     simply use [v], and in the latter, we should use [i]. But
+                     since we can't the tell the cases apart here, we'll just
+                     return the non empty list *)
+                  v &&& i >>^ (function
+                    | [], modules
+                    | modules, [] -> modules
+                    | (_ :: _ as vlib), (_ :: _ as impl) ->
```

Comment:
the _::_ patterns shouldn't be required here. If you're keeping them for documentation it's OK, otherwise that can be simplified to `(vlib, impl)`

Reply:
Yup, it's only for documentation.